plugins/SpectrumAnalyzer/Analyzer.cpp

```diff
@@ -0,0 +1,75 @@
+/*
+ * Analyzer.cpp - definition of Analyzer class.
+ *
+ * Copyright (c) 2019 Martin Pavelek <he29/dot/HS/at/gmail/dot/com>
+ *
+ * Based partially on Eq plugin code,
+ * Copyright (c) 2014-2017, David French <dave/dot/french3/at/googlemail/dot/com>
+ *
+ * This file is part of LMMS - https://lmms.io
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#include "Analyzer.h"
+
+#include "embed.h"
+#include "plugin_export.h"
+
+
+extern "C" {
+	Plugin::Descriptor PLUGIN_EXPORT analyzer_plugin_descriptor =
+	{
+		STRINGIFY(PLUGIN_NAME),
+		"Spectrum Analyzer",
+		QT_TRANSLATE_NOOP("pluginBrowser", "A graphical spectrum analyzer."),
+		"Martin Pavelek <he29/dot/HS/at/gmail/dot/com>",
+		0x0100,
+		Plugin::Effect,
+		new PluginPixmapLoader("logo"),
+		NULL,
+		NULL
+	};
+}
+
+
+Analyzer::Analyzer(Model *parent, const Plugin::Descriptor::SubPluginFeatures::Key *key) :
+	Effect(&analyzer_plugin_descriptor, parent, key),
+	m_controls(this),
+	m_processor(&m_controls)
+{
+}
+
+
+// Take audio data and pass them to the spectrum processor.
+// Skip processing if the controls dialog isn't visible, it would only waste CPU cycles.
+bool Analyzer::processAudioBuffer(sampleFrame *buffer, const fpp_t frame_count)
+{
+	if (!isEnabled() || !isRunning ()) {return false;}
+	if (m_controls.isViewVisible()) {m_processor.analyse(buffer, frame_count);}
+	return isRunning();
```

Comment:
Umm, I am not quite sure what is checkGate() supposed to do? From comment in the function, I assume it stops (all?) processing when the output of all (or one?) effects becomes zero, or something?
Since the analyzer does not change the audio buffer in any way, there should be no reason to report back to LMMS, since the result would be the same as from the previous effect, right?

Btw., I literally just assumed the EQ plugin does everything "the right way" and did most of the overall structure and LMMS interface the same way, so there may be thing like this that are completely new and unknown to me.. :))

Reply:
You are right that it somehow informs LMMS whether the output shall be re-used. But I think it also has [another function](https://github.com/JohannesLorenz/lmms/blob/lv2-review/src/core/Effect.cpp#L150-L154). It seems to stop your effect from running when the output is zero for some timeout. So you could use isRunning() to check whether the analyzer should proceed (many of our plugins check it). Does this make sense, or am I reading it wrong?

Reply:
I'm still not sure if I understand it correctly. :) So essentially, calling checkGate() provides LMMS one of the inputs that it uses to determine if a given effect should run or not? I already check it (the first line here, just return false of not running or not enabled), so that part should be all right.
What I don't like is the timeout thing: the waterfall history is generated in the analyse() method, so if a relatively quick timeout stopped the plugin, the currently displayed waterfall would just stop "falling down".

Reply:
I just tried adding the code you referenced in the first comment and it simply stops the plugin as soon as I stop playing the default TripleOsc tone, freezing both the waterfall and the spectrum display in place. So unless there is way to disable the "empty" check, only keep the timeout and increase the timeout value so that it does not freeze too soon, this would simply break the plugin.

Reply:
OK, it should not freeze. It should rather do no new computations and the remaining stuff should just flow out until the waterfall diagram is empty. Then, it can stop.

I wonder if this `checkGate` is necessary. In most cases, I'd say this is cosmetic, but FWIR the plugin can take a lot of CPU power. On the other hand, I observe that CPU usage goes down as soon as I stop playing a note... Is there anything in the code bypassing computations when the input is silence? I couldn't find it, but my CPU says there is?

Reply:
I think checkGate simply does not make sense here, because whether there is any non-zero output from the plugin or not does not really tell you much about whether the plugin can be stopped without causing any issues.
The FFT processing runs only when either the spectrum or the waterfall are displayed. It could be optimized by checking if all input samples are zero and filling result buffers with zeros instead of running FFT, but since fftw runs pretty fast, I'm not sure if the potential gains are worth the complications in code. They may be.
The waterfall view is a "dumb image painter", i.e. it knows nothing about the content and draws it as it finds it (if the widget is displayed). But the scaling and drawing is very fast and efficient (probably well optimized in Qt) so the CPU footprint of having it run even if it is empty should be minimal (at least for block sizes <= 4096).
The main CPU hog and performance limiting widget is the spectrum view: QPainter is quite slow when drawing a lot of small line segments. I even tried converting it into an OpenGL widget and QtQuick widget (singe QtQuick should be accelerated by default) and nothing changed (only for worse). So I added some quality / resolution controls to make sure the curve does not have "too much detail", and just as the Equalizer, I stop the drawing when all displayed values reach zero, which is probably the main reason CPU usage drops quickly after you release a key.

Reply:
Ah if it's `QPainter`, that explains why the CPU usage goes down after releasing the keys. I thought it was the FFT. In this case, I don't see a reason to call `checkGate()` either. I'm hitting "resolve".